source/extensions/stat_sinks/hystrix/hystrix.cc

```diff
@@ -0,0 +1,375 @@
+  } else {
+    rolling_stats_map_[key][current_index_] = value;
+  }
+}
+
+uint64_t HystrixStatCache::getRollingValue(absl::string_view cluster_name, absl::string_view stat) {
+  std::string key;
+  key = absl::StrCat("cluster.", cluster_name, ".", stat);
+  if (rolling_stats_map_.find(key) != rolling_stats_map_.end()) {
+    // If the counter was reset, the result is negative
+    // better return 0, will be back to normal once one rolling window passes.
+    if (rolling_stats_map_[key][current_index_] <
+        rolling_stats_map_[key][(current_index_ + 1) % window_size_]) {
+      return 0;
+    } else {
+      return rolling_stats_map_[key][current_index_] -
+             rolling_stats_map_[key][(current_index_ + 1) % window_size_];
+    }
+  } else {
+    return 0;
+  }
+}
+
+void HystrixStatCache::CreateCounterNameLookupForCluster(const std::string& cluster_name) {
+  // Building lookup name map for all specific cluster values.
+  // Every call to the updateRollingWindowMap function should get the appropriate name from the map.
+  std::string cluster_name_with_prefix = absl::StrCat("cluster.", cluster_name, ".");
+  counter_name_lookup[cluster_name]["upstream_rq_timeout"] =
+      absl::StrCat(cluster_name_with_prefix, "upstream_rq_timeout");
+  counter_name_lookup[cluster_name]["upstream_rq_per_try_timeout"] =
+      absl::StrCat(cluster_name_with_prefix, "upstream_rq_per_try_timeout");
+  counter_name_lookup[cluster_name]["timeouts"] =
+      absl::StrCat(cluster_name_with_prefix, "timeouts");
+  counter_name_lookup[cluster_name]["upstream_rq_5xx"] =
+      absl::StrCat(cluster_name_with_prefix, "upstream_rq_5xx");
+  counter_name_lookup[cluster_name]["retry.upstream_rq_5xx"] =
+      absl::StrCat(cluster_name_with_prefix, "retry.upstream_rq_5xx");
+  counter_name_lookup[cluster_name]["upstream_rq_4xx"] =
+      absl::StrCat(cluster_name_with_prefix, "upstream_rq_4xx");
+  counter_name_lookup[cluster_name]["retry.upstream_rq_4xx"] =
+      absl::StrCat(cluster_name_with_prefix, "retry.upstream_rq_4xx");
+  counter_name_lookup[cluster_name]["errors"] = absl::StrCat(cluster_name_with_prefix, "errors");
+  counter_name_lookup[cluster_name]["upstream_rq_2xx"] =
+      absl::StrCat(cluster_name_with_prefix, "upstream_rq_2xx");
+  counter_name_lookup[cluster_name]["success"] = absl::StrCat(cluster_name_with_prefix, "success");
+  counter_name_lookup[cluster_name]["upstream_rq_pending_overflow"] =
+      absl::StrCat(cluster_name_with_prefix, "upstream_rq_pending_overflow");
+  counter_name_lookup[cluster_name]["rejected"] =
+      absl::StrCat(cluster_name_with_prefix, "rejected");
+  counter_name_lookup[cluster_name]["total"] = absl::StrCat(cluster_name_with_prefix, "total");
+}
+
+void HystrixStatCache::updateRollingWindowMap(std::map<std::string, uint64_t> current_stat_values,
+                                              std::string cluster_name) {
+
+  if (counter_name_lookup.find(cluster_name) == counter_name_lookup.end()) {
```

Comment:
ditto -- lots of lookups to cluster_name all of which return the same exact string->string map.